### ISSUE TITLE
Design proposal for `/itchysats/rollover` protocol

### ISSUE THREAD

**Issue by thomaseizinger** [itchysats/itchysats#1775]
## Context

Here is what we know about the current rollover protocol:

1. We have an instance of the [two-generals problem](https://en.wikipedia.org/wiki/Two_Generals%27_Problem). Either party should only move to the new state if the other party completed the protocol successfully. We tried to naively solve this with ACKs but literature tells us that this doesn't do anything.
2. During rollover, the parties share revocation keys. Once a revocation key is shared, we have to assume it can be used and thus we MUST NOT go back to a previous commit transaction.
3. The current protocol does not explicitly state, which commit transaction we are rolling over over from. This creates problems in case we run into a 2-general situation where one party completes successfully and they other party does not.

## New protocol

The design proposal for a new protocol is as follows:

```mermaid
sequenceDiagram
	Dialer->>Listener: Open substream `/itchysats/rollover`
	Note right of Listener: Contains CfdId, current commit transaction and `BitMexPriceEventId`
	Listener->>Dialer: Send Accept
	Note right of Listener: Reject on unknown CFD, commit transaction or bad event ID
	Dialer->>Listener: Send PubKeys
	Listener->>Dialer: Send PubKeys
	Dialer->>Listener: Send Signatures
	Listener->>Dialer: Send Signatures
	Dialer->>Listener: Send SecretKey
	Listener->>Dialer: Send SecretKey
	Dialer->>Listener: Send Done
```

```rust
struct Accept {
	funding_rate: f64
}

struct PubKeys {
	revocation: Bytes,
	publish: Bytes,
}

struct Signatures {
    commit: EcdsaAdaptorSignature,
    cets: HashMap<String, Vec<(RangeInclusive<u64>, EcdsaAdaptorSignature)>>,
    refund: Signature,
}

struct SecretKey {
	revocation: Bytes
}
```

The pillars of the new design are:

1. Sending the current commit transaction and event ID allows us to detect state disparity
2. Send messages one after the other to get implicit ACKs
3. Send final ACK to know whether dialer succeeded
4. Listener is idempotent IF previous rollover failed (no ACK received)

## Detailed explanation

The two-generals problem tells us that it is not possible for both parties to know whether they other party successfully completed the protocol. However, what is possible is for the listener to know, whether the dialer succeeded if we add a final `Done` message. If the listener receives this, they know that the dialer completed successfully.

Additionally, by including the commitment transaction we are rolling over from in the initial message, we can design the listener in an idempotent way: We can allow the dialer to re-execute the rollover by using the exact same parameters. To ensure this, we include the `funding_rate` in the `Accept` message. This allows us to send the `funding_rate` that was used in the original, partially failed rollover and thus computing the exact same rollover again.
Doing so is important because a malicious dialer might re-trigger rollover even though they actually succeeded.

A key thing to note about the idempotent design is that we need only allow "re-rollovers" for previously failed ones. This is the motivation for including the `Done` message. Without this restriction, a malicious party could perform a DoS attack by constantly re-triggering rollover for the same CFD.

In case the dialer fails after sharing the revocation key, we must remember to not use the old commit transaction. In theory, a malicious listener may purposely abort the protocol and deny any attempt of retrying the rollover protocol. This leaves the dialer in a state where they cannot use the old commit transaction but the listener can without the dialer being able to revoke that state. I don't think this situation can be solved elegantly at the protocol level. Once relevant, we should aim to detect such behaviour as "bad intent" and not interact with that peer.

**Comment by da-kami** [comment]
Some general notes:

- If the listener did not receive the `ack` from the dialer, but the dialer (for whatever reason) still believes it was sent successfully the listener has to allow the dialer to trigger a *next* rollover;  i.e. the listener should not enforce a retry by the dialer!
In such a case the rollover state will match and the next rollover will succeed.
- The change to include the `event_id` as part of the proposal is actually independent of making the protocol resilient to failures at the end of the execution (i.e. the `ack` step). I propose to remove this change and re-evaluate it later. The current design where the maker dictates the fees is resilient.
- Similarly, we don't have to include the `commit_tx` in the proposal because it only helps to fail earlier, but does not aid resilience against failure.

Based on the proposal in the description I propose these changes to the current `itchysats/rollover` protocol:

1. ~~Include the current `commit_tx` and `event_id` in the proposal.~~
    1. ~~The listener rejects proposal if current `commit_tx` or the proposed `event_id` don't match~~
2. Only the dialer sends `Msg3` (= `ack`), remove sending `Msg3` from the listener - *breaking protocol change*
3. The listener records if the dialer's `ack` was received at the end of the protocol. If the ack was not received the listener has to record additional state in (or in addition to) `RolloverCompleted`:
    1. The generated random keys used during protocol execution
    2. The used `tx_fee_rate`
    3. The previous `commit_tx` (can be know from succeeded rollover's list of revoked commits, i.e. the last one in the list)
    4. The `current_fee` based on the used offer's funding rate (can be know from succeeded rollover)
    5. The `event_id` used (can be know from succeeded rollover)
5. If the dialer fails to send `Msg3` after sending the keys the dialer has to record additional state in (or in addition to) `RolloverFailed`:
    1. The generated random keys used during protocol execution
    2. The used `tx_fee_rate`
    4. The `current_fee` based on the used offer's funding rate (the listener can know this from the succeeded rollover)
    6. The `event_id` used (the listener can know this from the succeeded rollover)
7. The proposal message from dialer to listener needs to be extended with an optional parameter that marks it as `retry` so that the listener knows how to handle it. 
    1. If the last rollover by the listener has a missing `ack` retry will be allowed, otherwise not.
    2. The listener has to use the previous parameters to evaluate the proposal. For a retry the `commit_tx` and `event_id` will have to be taken from the state that was saved for retry upon missing `ack`
8. If the dialer triggers a rollover retry, and the retry is eligible on the listener side, the listener and dialer will have to use the previously used keys for the protocol execution.
   1. At the end of the protocol execution the listener should verify that the outcome is the same, i.e. the saved rollover equals the result of the retry. The listener does not save the successfully retried rollover DLC, because the previous one is already valid and active. 
   2. Given that the listener received the dialer's `ack` from the retried rollover the listener has to update the state to record that the ack was received. No further retries will be allowed then.

**Comment by da-kami** [comment]
After discussion with @bonomat and @klochowicz we came to the conclusion that we can achieve a retry mechanism in a different that is more effective because it will not require any `ack` messages.

Message design:

```mermaid
sequenceDiagram
	Dialer->>Listener: Open substream `/itchysats/rollover`
	Dialer->>Listener: Proposal [commit_tx]
        Listener->>Listener: When was this commit_tx? <br>Calculate costs from that commit_tx
        Listener->>Dialer: Accept[These are the costs]
        Dialer->>Listener: Send PubKeys
	Listener->>Dialer: Send PubKeys
	Dialer->>Listener: Send Signatures
	Listener->>Dialer: Send Signatures
	Dialer->>Listener: Send SecretKey
	Listener->>Dialer: Send SecretKey
        Note right of Listener: Fire and forget send <br>last message  <br>(ensure we complete)
        Listener->>Listener: Append Completed
       Note left of Dialer: It is irrelevant <br>if the dialer  <br>succeeds or not
```

Essentials:

1. In the proposal the dialer (taker) adds the `commit_tx` from which to rollover.
2. The listener (maker) uses the `commit_tx` to decide from when the taker wants to roll over. 
    1.  The listener calculates the fees retrospectively, i.e. all the fees from
    2. If the listener does not know the `commit_tx` that the dialer proposes the proposal will be rejected. 
    This should never happen because the listener is the last one sending and we fire and forget.
4. The listener communicates the fees in Accept (this is already implemented, but might need some adaption)
5. The listener is the one sending the last message as fire and forget to ensure that it completes.
6. If the dialer fails but the listner succeeds the dialer will just propose from the same `commit_tx` know to the dialer. The listener knows all previous `commit_tx`. 

Additional rules:

7. The dialer can only propose a `commit_tx` for which the `settlement_event_id` is less than `settlement_interval` (i.e. 24h) time in the past. **To be able to decide eligibility the maker will have to store the `settlement_event_id` for the `RevokeCommit`**

**Comment by klochowicz** [comment]
This will also close: https://github.com/itchysats/itchysats/issues/1622

**Comment by da-kami** [comment]
Once we started implement this we ran into one obstacle:

At the moment the `RolloverCompleted` event `funding_fee` is still accumulative, i.e. for every rollover in the database when building the `model::Cfd` aggregate we accumulate the fee.

Now, with our change it can happen that the maker performs multiple rollovers for the same `settlement_event_id` for one `Cfd`. But, the fee should not  be accumulative then; only the *latest* fee should apply. 
We cannot just remove the accumulated fees (i.e. we cannot just remove the mechanism of the `FeeAccount`) because the maker will have to be able to come up with a "start fee" for historical rollover requests.

There are different ways to overcome this:

1. Keep the `funding_fee` accumulative but add decision logic to the aggregate that makes sure to only apply the latest fee if we notice we have `RolloverCompleted` for the same `event_id`.
2. Only store the complete fee for a rollover and overwrite the `funding_fee` in the `fee_account` when building the CFD from events instead of accumulating it. Additionally, store the `complete_fee` for each `RevokeCommit` so we can "start" from there again.

I would try to move forward with (2), given that there is only one "active" rollover in the system now, I think the accumulated `funding_fee` is super error prone anyway.

**Comment by klochowicz** [comment]
I don't think only complete fee will work; you will likely run into problems of not being able to do any retries with that. 
I will create a POC with `funding_fee` snapshots for all the rollovers that happened, so that we can accurately send back a complete fee for all possible cases (so that maker and taker can reconcile back if they diverge).

**Comment by da-kami** [comment]
> I don't think only complete fee will work; you will likely run into problems of not being able to do any retries with that.

Why? 
I think it would be more resilient to use the `complete_fee` instead of assembling it every time. The `complete_fee` is what is actually used for calculating the payouts in the rollover and is agreed upon between the two parties when you roll over.

I stated

> Additionally, store the complete_fee for each RevokeCommit so we can "start" from there again.

This is what would allow going from any point in time, because the maker would know what the fee was for a specific `commit_tx`. 

1. The taker sends a rollover proposal for `commit_tx` x
2. The maker goes through the list of `commit_tx` and finds x
3. The maker validates that the `event_id` attached to `commit_tx` x is still within the range of 24h - see https://github.com/itchysats/itchysats/issues/1775#issuecomment-1162610819
4. The maker picks the fee attached to `commit_tx` x as base fee.
5. The maker calculates the fee to be charged based on the base fee + time between `commit_tx`-`event_id` and the `event_id` that will be rolled over to.

I think this works and is more resilient than building up the fee on both sides.
When we apply the event we then set the complete fee additionally. For backwards compatibility we make the `complete_fee` optional and either `set` it (if available, this will just overwrite the complete fee in the `fee_account`) or (if complete fee is `None`) we accumulate with the old code.

On the taker side this will work because the parties agree on the complete fees during rollover, so this would always be aligned.
It is backwards compatible as long as the taker does not trigger a rollover for the same event-id (but for such kind of cases we would run into problems with the current implementation as well).

**Comment by da-kami** [comment]
@klochowicz see https://github.com/itchysats/itchysats/pull/2272/commits/ecd142e7aca495bc4cedf2e79c5da828af228b44
where I am using the `complete_fee`.
I opted for keeping he `FeeAccount` for now and just initialize it from the `complete_fee`. I think this is the better solution then trying to "filter out" the same event-id payments. this will add more resilience overall because we stop accumulating on both sides.

**Comment by da-kami** [comment]
**TLDR;**

After a quick discussion with @bonomat we agreed that the feature described below can be done in a follow up if this requires extensive changes in the tests. I will have a quick look once done with the database changes and either apply a qick fix to tests or open a follow up ticket.

---

**Details**

The maker actually does not restrict rollovers to `event-id`s that have not occurred yet.
I think we did that on purpose - because the maker ensures to commit anyway, so we just did not restrict rollover. 
This made our rollover actor tests (and potentially other tests...) simpler, because we can use an `event-id` that from the past - that's why we always see this message in the logs when running the rollover actor tests:

```
2022-06-22T03:24:35.319244Z  WARN model::cfd: Rolling over a contract that can be settled non-collaboratively
```

With the new rollover protocol the maker allows the taker to roll over from an event-id that was already used.
When it comes to the maker validating the taker's request I stated that the maker should validate that the `event-id` related to the `commit-tx` proposed by the taker is still valid, i.e.

> 3. The maker validates that the event_id attached to commit_tx x is still within the range of 24h

But this would cause problems with the tests, because we would now introduce the restriction that we previously (on purpose?) avoided.
I think it would be better to implement a restriction to `event-id`s within `24h` in the new rollover. This can quite easily achieved by using `has_likely_occurred`. Test changes have to be evaluated.

**Comment by da-kami** [comment]
Closing this because #2272 merged which implements the protocol proposed in https://github.com/itchysats/itchysats/issues/1775#issuecomment-1160081024
